sky/utils/plugin_extensions/external_failure_source.py

```diff
@@ -0,0 +1,172 @@
+"""External failure source interface for plugins.
+
+This module provides an extension point that allows plugins to provide
+cluster failure tracking functionality. By default, no-op implementations
+are used. Plugins can register their own implementations to provide actual
+failure tracking.
+
+Example usage in a plugin:
+    from sky.utils.plugin_extensions import ExternalFailureSource
+
+    # Register custom failure source
+    ExternalFailureSource.register(
+        get_failures=my_get_cluster_failures,
+        clear_failures=my_clear_cluster_failures,
+    )
+
+Example usage in core SkyPilot:
+    from sky.utils.plugin_extensions import ExternalFailureSource
+
+    # Get failures for a cluster
+    failures = ExternalFailureSource.get(cluster_hash='abc123')
+
+    # Clear failures for a cluster
+    cleared = ExternalFailureSource.clear(cluster_name='my-cluster')
+"""
+import dataclasses
+from typing import Any, Dict, List, Optional, Protocol
+
+from sky import sky_logging
+
+logger = sky_logging.init_logger(__name__)
+
+
+@dataclasses.dataclass
+class ExternalClusterFailures:
+    """Represents a cluster failure from an external source (e.g., GPU Healer).
+
+    Attributes:
+        code: Machine-readable failure code (e.g. 'GPU_HARDWARE_FAILURE_XID_79')
+        reason: Human-readable description of the failure.
+    """
+    code: Optional[str] = None
+    reason: Optional[str] = None
+
+    @classmethod
+    def from_failure_list(
+            cls, failures: List[Dict[str, Any]]) -> 'ExternalClusterFailures':
+        """Create from a list of failure dicts returned by
+        ExternalFailureSource.
+
+        Args:
+            failures: List of dicts with 'failure_mode' and 'failure_reason'
+            keys.
+
+        Returns:
+            ExternalClusterFailures with codes/reasons joined by '; '.
+        """
+        if not failures:
+            return cls()
+        codes = [f['failure_mode'] for f in failures]
+        reasons = [f['failure_reason'] for f in failures]
+        return cls(code='; '.join(codes), reason='; '.join(reasons))
```

Comment:
I feel this is not the right data structure, for multiple failures it should be:
list(
  ExternalClusterFailure(code=xx, reason=yy),
  ExternalClusterFailure(code=aa, reason=bb)
)